google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenTelemetryApiTracer.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.api.gax.tracing.ApiTracer;
+import com.google.api.gax.tracing.ApiTracerFactory.OperationType;
+import com.google.common.base.Preconditions;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+/**
+ * {@link com.google.api.gax.tracing.ApiTracer} for use with OpenTelemetry. Based on {@link
+ * com.google.api.gax.tracing.OpencensusTracer}.
+ */
+class OpenTelemetryApiTracer implements ApiTracer {
```

Comment:
We decided to separate the logic to calculate the metrics and the logic to record the metrics while implementing metrics in Gax. Hence [MetricsTracer](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsTracer.java) and [MetricsRecorder](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java) are generic classes without any knowledge of OpenTelemetry, only [OpenTelemetryMetricsRecorder](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/tracing/OpenTelemetryMetricsRecorder.java) is dependent on OpenTelemetry apis.
Spanner does not have to follow the same design principle and it may not be feasible for traces either, but something to consider.

Reply:
The Spanner client supports tracing with both OpenCensus and OpenTelemetry. The customer makes a choice at startup, and then the client library configures itself accordingly. We want to support the same for the `ApiTracer`, so based on that, it would make sense to create a technology-agnostic `ApiTracer`. However, there is already an [OpencensusTracer](https://github.com/googleapis/sdk-platform-java/blob/5799827a864be627bac03969cc178efc9d6170aa/gax-java/gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java#L210) in gax that can be used with Spanner today (and that we need to continue to support). That means that it is easier to create an `OpenTelemetry` sibling of this tracer, and choose one or the other based on the user configuration when the client is created, instead of adding another layer of abstraction between the client library and the `ApiTracer`.